core/common/src/main/java/alluxio/collections/UnmodifiableArrayList.java

```diff
@@ -0,0 +1,311 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.collections;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterators;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.util.Collection;
+import java.util.ConcurrentModificationException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+/**
+ * This class provides an unmodifiable List proxy for an underlying array.
+ * In other words, this helps you use the array like a List, but not being able to modify it.
+ *
+ * Create an instance with the constructor and get the underlying array with {@link #toArray()}.
```

Comment:
If we want to totally disable modification, the `toArray()` can be removed. But that may make the interface less handy.

Reply:
I think this interface just intends to guard accidental modifications, instead of anything really like safety.